server/aesdsocket.c

```diff
@@ -0,0 +1,366 @@
+
+    if( tmp_fd == -1 )
+    {
+        syslog(LOG_ERR, "Error opening file with error %d", errno);
+    }
+
+    char buffer[BUFFER_SIZE];
+    int byte_recv;
+    int packet_len;
+
+    bool data_packet = false;
+    int file_size = 0;
+
+    while(1)
+    {
+        new_fd = accept(sockfd, (struct sockaddr *)&their_addr, &addr_size);
+        if( new_fd == -1 )
+        {
+            syslog(LOG_ERR, "Error : accept with error no : %d", errno);
+            // accept new connection
+            continue;
+        }
+
+        // Reference : https://man7.org/linux/man-pages/man3/inet_ntop.3.html
+        char address_string[INET_ADDRSTRLEN];
+        struct sockaddr_in *p = (struct sockaddr_in *)&their_addr;
+        syslog(LOG_DEBUG, "Accepted connection from %s", inet_ntop(AF_INET, &p->sin_addr, address_string, sizeof(address_string)));
+
+        //clearing old values
+        memset(buffer, '\0', BUFFER_SIZE);
+        int cycle_count = 0;
+        int total_len = 0;
+        int old_size = BUFFER_SIZE;
+        data_packet = false;
+
+        while(!data_packet)
+        {
+            packet_len = 0;
+
+            if( (byte_recv =  recv(new_fd, &buffer, BUFFER_SIZE, 0)) == -1 )
+            {
+                syslog(LOG_ERR, "recv");
+            }
+            for(int i=0; i< BUFFER_SIZE; i++)
+            {
+                packet_len++;
+                if( buffer[i] == '\n' )
+                {
+                    data_packet = true;
+                    break;
+                }
+            }
+
+            if(cycle_count == 0)
+            {
+                recieve_buffer = (char *)malloc(packet_len);
+                if(recieve_buffer == NULL)
+                {
+                    syslog(LOG_ERR, "malloc");
+                }
+                // clearing buffer
+                memset(recieve_buffer, '\0', packet_len);
+                total_len = packet_len;
+            }
+            else
+            {
+                char *ptr = realloc(recieve_buffer, old_size+packet_len);
+                if(ptr == NULL)
+                {
+                    syslog(LOG_ERR, "realloc");
+                    
+                }
+                else
+                {
+                    recieve_buffer = ptr;
+                    old_size += packet_len;
+                }
+                total_len = old_size;
+            }
+        memcpy(recieve_buffer + (cycle_count * BUFFER_SIZE), buffer, packet_len);
+        cycle_count ++;
+        }
+        // move to end of file
+        lseek(tmp_fd, 0, SEEK_END);
+        int wc = write(tmp_fd, recieve_buffer, total_len);
+        if( wc == -1)
+        {
+            syslog(LOG_ERR, "write");
+        }
+        file_size += wc;
+        free(recieve_buffer);
+        // resend data to client
+        memset(buffer, '\0', BUFFER_SIZE);
+        lseek(tmp_fd, 0, SEEK_SET);
+        int read_bytes;
+
+        char *read_buffer = (char *)malloc(file_size);
```

Comment:
@MaanasMDK what assumption are you making here about the size of the process heap and the total size of all packets written?

Reply:
I believe I was more focused on the test program. This would not scale for extremely large packet sizes.